rspotify-model/src/idtypes.rs

```diff
@@ -296,9 +224,311 @@ impl<T: IdType> Id<T> {
+    /// `show`, or `episode`, and `{id}` is a non-empty alphanumeric string.
+    /// The URI must be match with the ID's type (`Id::TYPE`), otherwise
+    /// `IdError::InvalidType` error is returned.
+    ///
+    /// Examples: `spotify:album:6IcGNaXFRf5Y1jc7QsE9O2`,
+    /// `spotify/track/4y4VO05kYgUTo2bzbox1an`.
+    ///
+    /// If input string is not a valid Spotify URI (it's not started with
+    /// `spotify:` or `spotify/`), it must be a valid Spotify object id,
+    /// i.e. a non-empty alphanumeric string.
+    ///
+    /// # Errors:
+    ///
+    /// - `IdError::InvalidType` - if `id_or_uri` is an URI, and it's type part
+    ///    is not equal to `T`,
+    /// - `IdError::InvalidId` - either if `id_or_uri` is an URI with invalid id
+    ///    part, or it's an invalid id (id is invalid if it contains
+    ///    non-alphanumeric characters),
+    /// - `IdError::InvalidFormat` - if `id_or_uri` is an URI, and it can't be
+    ///    split into type and id parts.
+    fn from_id_or_uri(id_or_uri: &str) -> Result<Self, IdError>
+    where
+        Self: Sized,
+    {
+        match Self::from_uri(id_or_uri) {
+            Ok(id) => Ok(id),
+            Err(IdError::InvalidPrefix) => Self::from_id(id_or_uri),
+            Err(error) => Err(error),
+        }
+    }
+}
+
+pub trait PlayableId: Id {}
+pub trait PlayContextId: Id {}
+
+/// This macro helps consistently define ID types. It contains a lot of code but
+/// mostly it's just repetitive work that's not of much interest for being
+/// trivial.
+///
+/// * The `$type` parameter indicates what type the ID is made out of (say,
+///   `Artist`, `Album`...) from the enum `Type`.
+/// * The `$name` parameter is the identifier of the struct for that type.
+macro_rules! define_idtypes {
+    ($($type:ident => $name:ident),+) => {
+        $(
+            #[doc = concat!(
+                "ID of type [`Type::",
+                stringify!($type),
+                "`]. Refer to the [module-level docs][`crate::idtypes`] for more information.")
+            ]
+            #[derive(Clone, Debug, PartialEq, Eq, Serialize, Hash)]
+            pub struct $name(String);
+
+            impl Id for $name {
+                #[inline]
+                fn id(&self) -> &str {
+                    &self.0
+                }
+
+                #[inline]
+                fn _type(&self) -> Type {
+                    Type::$type
+                }
+
+                #[inline]
+                fn _type_static() -> Type where Self: Sized {
+                    Type::$type
+                }
+
+                #[inline]
+                unsafe fn from_id_unchecked(id: &str) -> Self {
+                    $name(id.to_owned())
+                }
+            }
+
+            // Deserialization may take either an Id or an URI, so its
+            // implementation has to be done manually.
+            impl<'de> Deserialize<'de> for $name {
```

Comment:
I am not sure if it's necessary to implement the `serialize` function manually? Otherwise, the serialized output will be different from the original `Json` data?

Reply:
The thing is that when deserializing we want to be more flexible and accept both IDs and URIs. And when serializing we can only choose one, so IDs make the most sense to me.

Though I think the problem here is that these models shouldn't even implement `Serialize`. At no point in our code do we need to serialize an ID and it's probably confusing to do so anyway. IMO we should remove the `#[derive(Serialize)]` from the IDs and many other types in the model with the same issue. I would leave that for another PR, though.